### ISSUE TITLE
pop_min and pop_max don't seem to be atomic when used from multiple threads

### ISSUE THREAD

**Issue by caelunshun** [spacejam/sled#1127]
`pop_min` and `pop_max` seem to yield duplicate keys when called from multiple threads.

```rust
use crossbeam::thread::scope;
use std::convert::TryInto;

let db = sled::open("db")?;

// Insert values 0..5
for x in 0u32..5 {
    db.insert(x.to_be_bytes(), &[])?;
}
 
// Pop 5 values using multiple threads
scope(|s| {
    for _ in 0..5 {
        s.spawn(|_| {
            let (key, _) = db.pop_min().unwrap().unwrap();
            let x = u32::from_be_bytes((&*key).try_into().unwrap());
            println!("{}", x);
        });
    }
})
.unwrap();

assert!(
    db.is_empty(),
    "elements left in database: {:?}",
     db.iter().collect::<Vec<_>>()
);
```

I would expect this example to print out the numbers 0..5 in some order, and for the database to be empty after performing 5 `pop_min` operations. However, for some reason `pop_min` seems to yield the same key multiple times. Here's one output I got:
```
0
0
0
1
2
thread 'main' panicked at 'elements left in database: [Ok(([0, 0, 0, 3], [])), Ok(([0, 0, 0, 4], []))]', src/main.rs:22:5
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```
The same problem occurrs with `pop_max` instead of `pop_min`:
```
4
3
3
3
3
thread 'main' panicked at 'elements left in database: [Ok(([0, 0, 0, 0], [])), Ok(([0, 0, 0, 1], [])), Ok(([0, 0, 0, 2], []))]', src/main.rs:24:5
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

I've also tested this in `sled` versions from `0.32` back to `0.25` and the issue persists. Note that when calling `pop_min` sequentially instead of in parallel, this problem never arises.

For reference, I'm using `rustc` 1.44.1 on Linux with kernel version 4.15.

**Comment by spacejam** [comment]
Thank you for this bug report! This is quite a facepalm. I've added your provided code as a regression test to prevent this in the future. I've fixed the issue in #1128 and released 0.33 which includes the fix. 

This happened because of some confusion around the nested result type returned by Tree::compare_and_swap internally, which has actually helped to avoid many other bugs over time, but this is one rare case where it takes a couple back.

Thanks again for opening this, and please don't hesitate to open reports in the future if you encounter other issues!

**Comment by caelunshun** [comment]
Great, thank your for the quick resolution!
